### PR TITLE
feat: add caching layer to 360 images

### PR DESCRIPTION
**Pull request by christjt**
#### Type of change

Feat  

#### Jira ticket :blue_book:

https://cognitedata.atlassian.net/browse/REV-528

## Description :pencil:

Adds caching layer to 360 images. Currently we just keep loading new 360 images each time we click a new one which gets out of hand quite quickly. I've also added a preloading step which is triggered on hover which populates the cache.

## How has this been tested? :mag:

Unit tests :) 

## Checklist :ballot_box_with_check:

- [x] I am proud of this feature.
- [ ] I have performed a self-review of my own code.
- [x] I have added PR type (Feat, Bug, Chore, Test, Docs, Style, Refactor) to the PR title.
- [x] I have manually tested this for different scenarios (different models, formats, devices, browsers).
- [x] I have commented my code in hard-to-understand areas.
- [ ] I have made corresponding changes to the public documentation.
- [x] I have added unit and visuals tests to prove that my feature works to the best of my ability.
- [x] I have refactored the code for readability to the best of my ability.
- [ ] I have checked that my changes do not introduce regressions in the public documentation.
- [x] I have outlined any known defects / lacking capabilities in the contents of this PR.
- [x] I have listed any remaining work that should follow this PR in the description or jira/miro/etc.
- [ ] I have added TSDoc to any public facing changes.
- [ ] I have added "developer documentation" in any package README.md that is related / applicable to this PR.
- [x] I have noted down and am currently tracking any technical debt introduced in this PR.
- [x] I have thoroughly thought about the architecture of this implementation.
- [ ] I have asked peers to test this feature.

### CODE REVIEW

File: viewer/packages/360-images/src/Image360Entity.ts [hunk 1]
```diff
@@ -76,15 +79,50 @@ export class Image360Entity {
...
+    const materials =
+      imageContainerMaterial instanceof THREE.Material ? [imageContainerMaterial] : imageContainerMaterial;
+
+    materials
+      .map(material => material as THREE.MeshBasicMaterial)
+      .forEach(material => {
+        material.map?.dispose();
+        material.dispose();
+      });
+
```
Comment (haakonflatval-cognite) [inline viewer/packages/360-images/src/Image360Entity.ts:118]: Is this really the easiest way of disposing material? Would hope that the mesh did this automatically :/

File: viewer/packages/360-images/src/Image360Facade.ts [hunk 2]
```diff
@@ -21,10 +22,10 @@ export class Image360Facade<T> {
     this._image360Entities.forEach(entity => (entity.icon.hoverSpriteVisible = visible));
   }
 
-  constructor(entityFactory: Image360EntityFactory<T>) {
-    this._entityFactory = entityFactory;
+  constructor(private readonly _entityFactory: Image360EntityFactory<T>) {
```
Comment (haakonflatval-cognite) [inline viewer/packages/360-images/src/Image360Facade.ts:25]: So this `_entityFactory` argument is unneeded?

**Review by haakonflatval-cognite — APPROVED** [review]
Looks good! A few comments, questions

File: viewer/packages/360-images/src/Image360Entity.ts (hunk 1 above)
Comment (christjt) [reply viewer/packages/360-images/src/Image360Entity.ts:118]: It sadly cannot safely do this as two separate meshes can share the same material and there is no reference counting or any smart stuff going on afaik. Same goes for textures, as the texture can be in use in other materials etc.

File: viewer/packages/360-images/src/Image360Facade.ts (hunk 2 above)
Comment (christjt) [reply viewer/packages/360-images/src/Image360Facade.ts:25]: No, it is. Just syntactic sugar for setting it as a field. These two are equivilant:

```ts
private readonly _foo: Bar;
constructor(foo: Bar) {
   this._foo = foo;
  ...
}
```

```ts
constructor(private readonly _foo: Bar) {
  ...
}
```

**Pull request merged**